File: windows/src/Patchthrough.Core/AccuracyPipeline.cs
using System.Text.Json;

namespace Patchthrough.Core;

public sealed record ConfidenceCalibration(double Scale, double Offset)
{
    public double? Apply(double? confidence) => confidence is null
        ? null
        : Math.Clamp(confidence.Value * Scale + Offset, 0, 1);
}

public sealed record QualityProfile
{
    public sealed record EvidenceRecord
    {
        public required bool ReleaseQualified { get; init; }
        public double? DualEngineRelativeWerImprovement { get; init; }
        public bool? ConsensusNoCategoryRegression { get; init; }
    }

    public required string StandardEngine { get; init; }
    public required IReadOnlyList<string> MaxAccuracyEngines { get; init; }
    public required bool ConsensusQualified { get; init; }
    public required IReadOnlyDictionary<string, ConfidenceCalibration> Calibration { get; init; }
    public EvidenceRecord? Evidence { get; init; }

    public static QualityProfile SafeDefault { get; } = new()
    {
        StandardEngine = "parakeet",
        MaxAccuracyEngines = ["parakeet"],
        ConsensusQualified = false,
        Calibration = new Dictionary<string, ConfidenceCalibration>(),
    };

    public static QualityProfile Load(string? path = null)
    {
        path ??= Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
            ".config", "patchthrough", "quality-profile.json");
        if (!File.Exists(path)) return SafeDefault;
        try
        {
            return JsonSerializer.Deserialize<QualityProfile>(
                File.ReadAllText(path),
                new JsonSerializerOptions { PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower })
                ?? SafeDefault;
        }
        catch (JsonException)
        {
            return SafeDefault;
        }
    }

    public IReadOnlyList<string> Engines(string configured, QualityMode mode)
    {
        if (!string.Equals(configured, "auto", StringComparison.OrdinalIgnoreCase)) return [configured];
        if (Evidence?.ReleaseQualified != true) return ["parakeet"];
        if (mode == QualityMode.MaxAccuracy && CanRunConsensus && MaxAccuracyEngines.Count >= 2)
            return MaxAccuracyEngines.Take(2).ToList();
        return [mode == QualityMode.Standard ? StandardEngine : MaxAccuracyEngines.FirstOrDefault() ?? StandardEngine];
    }


    public bool CanRunConsensus => ConsensusQualified
        && Evidence?.DualEngineRelativeWerImprovement >= 0.10
        && Evidence?.ConsensusNoCategoryRegression == true;
}

public static class TranscriptConsensus
{
    private enum Direction { Both, Left, Right }
    private sealed record Step(Direction Direction, int Left, int Right);

    public static EngineTranscript Combine(
        EngineTranscript primary,
        EngineTranscript secondary,
        IReadOnlyDictionary<string, ConfidenceCalibration>? calibration = null)
    {
        calibration ??= new Dictionary<string, ConfidenceCalibration>();
        var words = new List<EngineWord>();
        foreach (var step in Align(primary.Words, secondary.Words))
        {
            if (step.Direction == Direction.Both)
            {
                var left = primary.Words[step.Left];
                var right = secondary.Words[step.Right];
                var leftScore = Score(primary.Engine, left.Confidence, calibration) ?? 0.5;
                var rightScore = Score(secondary.Engine, right.Confidence, calibration) ?? 0.5;
                var selected = leftScore >= rightScore ? left : right;
                words.Add(selected with { Confidence = Math.Max(leftScore, rightScore) });
            }
            else if (step.Direction == Direction.Left)
            {
                var word = primary.Words[step.Left];
                var score = Score(primary.Engine, word.Confidence, calibration);
                if (score is null || score >= 0.65) words.Add(word with { Confidence = score });
            }
            else
            {
                var word = secondary.Words[step.Right];
                var score = Score(secondary.Engine, word.Confidence, calibration);
                if (score >= 0.75) words.Add(word with { Confidence = score });
            }
        }
        words.Sort((a, b) => a.StartMs != b.StartMs
            ? a.StartMs.CompareTo(b.StartMs)
            : a.EndMs.CompareTo(b.EndMs));
        var text = Segmentation.Normalize(string.Join(" ", words.Select(word => word.Text)));
        var requested = OrderedUnion(primary.Context.RequestedTerms, secondary.Context.RequestedTerms);
        var applied = VocabularyEvidence.AcousticallySupported(requested, words);
        return new EngineTranscript
        {
            Engine = "rover",
            Model = primary.Model + "+" + secondary.Model,
            Version = "1",
            Settings = new Dictionary<string, string>
            {
                ["alignment"] = "timestamped-word-dp",
                ["engines"] = primary.Engine + "," + secondary.Engine,
            },
            Text = text,
            Language = primary.Language ?? secondary.Language,
            AudioDurationMs = Math.Max(primary.AudioDurationMs, secondary.AudioDurationMs),
            ProcessingDurationMs = primary.ProcessingDurationMs + secondary.ProcessingDurationMs,
            Words = words,
            Segments = Segmentation.From(words.Select(word => new WordTiming(
                word.Text, word.StartMs / 1000.0, word.EndMs / 1000.0, word.Confidence)).ToList()),
            Diagnostics = new Dictionary<string, string> { ["hypotheses"] = "2" },
            Context = new EngineContextEvidence(
                requested,
                OrderedUnion(primary.Context.DetectedTerms, secondary.Context.DetectedTerms),
                applied),
        };
    }

    private static IEnumerable<Step> Align(IReadOnlyList<EngineWord> left, IReadOnlyList<EngineWord> right)
    {
        var cost = new int[left.Count + 1, right.Count + 1];
        for (var i = 0; i <= left.Count; i++) cost[i, 0] = i;
        for (var j = 0; j <= right.Count; j++) cost[0, j] = j;
        for (var i = 1; i <= left.Count; i++)
        for (var j = 1; j <= right.Count; j++)
        {
            var substitution = cost[i - 1, j - 1] + PairCost(left[i - 1], right[j - 1]);
            cost[i, j] = Math.Min(substitution, Math.Min(cost[i - 1, j] + 1, cost[i, j - 1] + 1));
        }

        var steps = new List<Step>();
        var x = left.Count;
        var y = right.Count;
        while (x > 0 || y > 0)
        {
            if (x > 0 && y > 0)
            {
                var substitution = PairCost(left[x - 1], right[y - 1]);
                if (substitution < 2 && cost[x, y] == cost[x - 1, y - 1] + substitution)
                {
                    steps.Add(new Step(Direction.Both, --x, --y));
                    continue;
                }
            }
            if (x > 0 && cost[x, y] == cost[x - 1, y] + 1)
                steps.Add(new Step(Direction.Left, --x, -1));
            else
                steps.Add(new Step(Direction.Right, -1, --y));
        }
        steps.Reverse();
        return steps;
    }

    private static double? Score(
        string engine,
        double? confidence,
        IReadOnlyDictionary<string, ConfidenceCalibration> calibration) =>
        calibration.TryGetValue(engine, out var curve) ? curve.Apply(confidence) : confidence;

    private static bool Same(string left, string right) => Normalize(left) == Normalize(right);
    private static int PairCost(EngineWord left, EngineWord right)
    {
        var centers = Math.Abs((left.StartMs + left.EndMs) / 2 - (right.StartMs + right.EndMs) / 2);
        var separated = left.EndMs + 500 < right.StartMs || right.EndMs + 500 < left.StartMs;
        if (centers > 1_500 && separated) return 2;
        return Same(left.Text, right.Text) ? 0 : 1;
    }
    private static string Normalize(string word) => new(word.ToLowerInvariant().Where(char.IsLetterOrDigit).ToArray());

    private static List<string> OrderedUnion(IEnumerable<string> first, IEnumerable<string> second)
    {
        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        return first.Concat(second).Where(seen.Add).ToList();
    }
}

public static class EchoDedup
{
    private const int WindowMs = 1_500;

    public static List<Segment> DropMicEcho(IEnumerable<Segment> segments)
    {
        var all = segments.ToList();
        var system = all.Where(segment => segment.SourceTrack == "system").ToList();
        if (system.Count == 0) return all;
        return all.Where(mic =>
        {
            if (mic.SourceTrack != "mic" || mic.Words is null || mic.Words.Count < 3) return true;
            var nearby = system.Where(candidate =>
                candidate.EndMs + WindowMs >= mic.StartMs && candidate.StartMs - WindowMs <= mic.EndMs).ToList();
            var candidates = nearby.SelectMany(candidate => candidate.Words ?? []).ToList();
            var used = new HashSet<int>();
            var matches = 0;
            foreach (var word in mic.Words)
            {
                var index = -1;
                for (var candidateIndex = 0; candidateIndex < candidates.Count; candidateIndex++)
                {
                    var candidate = candidates[candidateIndex];
                    if (!used.Contains(candidateIndex)
                        && Math.Abs(candidate.StartMs - word.StartMs) <= WindowMs
                        && Same(candidate.Text, word.Text))
                    {
                        index = candidateIndex;
                        break;
                    }
                }
                if (index < 0) continue;
                used.Add(index);
                matches++;
            }
            var ratio = matches / (double)mic.Words.Count;
            var micConfidence = Mean(mic.Words.Select(word => word.Confidence)) ?? mic.Confidence ?? 0;
            var systemConfidence = Mean(used.Select(index => candidates[index].Confidence))
                ?? Mean(nearby.Select(segment => segment.Confidence)) ?? 0;
            return !(ratio >= 0.8 && systemConfidence >= micConfidence);
        }).ToList();
    }

    private static bool Same(string left, string right) => Normalize(left) == Normalize(right);
    private static string Normalize(string word) => new(word.ToLowerInvariant().Where(char.IsLetterOrDigit).ToArray());
    private static double? Mean(IEnumerable<double?> values)
    {
        var present = values.Where(value => value.HasValue).Select(value => value!.Value).ToList();
        return present.Count == 0 ? null : present.Average();
    }
}

public static class TranscriptCalibration
{
    public static EngineTranscript Apply(
        EngineTranscript transcript,
        IReadOnlyDictionary<string, ConfidenceCalibration> calibration)
    {
        if (!calibration.TryGetValue(transcript.Engine, out var curve)) return transcript;
        var diagnostics = transcript.Diagnostics.ToDictionary(pair => pair.Key, pair => pair.Value);
        diagnostics["confidence_calibration"] = "held-out-linear-v1";
        return transcript with
        {
            Words = transcript.Words.Select(word => word with { Confidence = curve.Apply(word.Confidence) }).ToList(),
            Segments = transcript.Segments.Select(segment => segment with
            {
                Confidence = curve.Apply(segment.Confidence),
                Words = segment.Words.Select(word => word with { Confidence = curve.Apply(word.Confidence) }).ToList(),
            }).ToList(),
            Diagnostics = diagnostics,
        };
    }
}

File: Sources/patchthrough/Transcription/AccuracyPipeline.swift
import Foundation

struct QualityProfile: Codable, Sendable {
    struct Calibration: Codable, Sendable {
        let scale: Double
        let offset: Double

        func apply(_ confidence: Double?) -> Double? {
            confidence.map { min(max($0 * scale + offset, 0), 1) }
        }
    }

    struct Evidence: Codable, Sendable {
        let releaseQualified: Bool
        let dualEngineRelativeWerImprovement: Double?
        let consensusNoCategoryRegression: Bool?
    }

    let standardEngine: String
    let maxAccuracyEngines: [String]
    let consensusQualified: Bool
    let calibration: [String: Calibration]
    let evidence: Evidence?

    static let safeDefault = QualityProfile(
        standardEngine: "parakeet",
        maxAccuracyEngines: ["parakeet"],
        consensusQualified: false,
        calibration: [:],
        evidence: nil
    )

    static func load() -> QualityProfile {
        let url = FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent(".config/patchthrough/quality-profile.json")
        guard let data = try? Data(contentsOf: url) else { return .safeDefault }
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        return (try? decoder.decode(QualityProfile.self, from: data)) ?? .safeDefault
    }

    func engines(configured: String, mode: QualityMode) -> [String] {
        if configured != "auto" { return [configured] }
        guard isReleaseQualified else { return ["parakeet"] }
        if mode == .maxAccuracy, canRunConsensus, maxAccuracyEngines.count >= 2 {
            return Array(maxAccuracyEngines.prefix(2))
        }
        return [mode == .standard ? standardEngine : (maxAccuracyEngines.first ?? standardEngine)]
    }

    /// Product UI may offer Max Accuracy only after corrected-corpus evidence
    /// qualifies the profile. Until then, selecting it would be a cosmetic
    /// switch over the same recoverable Parakeet baseline.
    var maxAccuracyAvailable: Bool {
        isReleaseQualified && !maxAccuracyEngines.isEmpty
    }

    var isReleaseQualified: Bool {
        evidence?.releaseQualified == true
    }

    var canRunConsensus: Bool {
        consensusQualified
            && (evidence?.dualEngineRelativeWerImprovement ?? 0) >= 0.10
            && evidence?.consensusNoCategoryRegression == true
    }
}

enum TranscriptConsensus {
    private enum Step { case both(Int, Int), left(Int), right(Int) }

    /// Two-engine ROVER: align words, retain agreements, and resolve real
    /// disagreements with calibrated confidence. Timing is never invented.
    static func combine(
        _ primary: EngineTranscript,
        _ secondary: EngineTranscript,
        calibration: [String: QualityProfile.Calibration] = [:]
    ) -> EngineTranscript {
        let aligned = align(primary.words, secondary.words)
        var words: [TimedWord] = []
        for step in aligned {
            switch step {
            case .both(let left, let right):
                let a = primary.words[left], b = secondary.words[right]
                let aScore = calibration[primary.engine]?.apply(a.confidence) ?? a.confidence ?? 0.5
                let bScore = calibration[secondary.engine]?.apply(b.confidence) ?? b.confidence ?? 0.5
                words.append(aScore >= bScore ? calibrated(a, aScore) : calibrated(b, bScore))
            case .left(let index):
                let word = primary.words[index]
                let score = calibration[primary.engine]?.apply(word.confidence) ?? word.confidence
                if score == nil || score! >= 0.65 { words.append(calibrated(word, score)) }
            case .right(let index):
                let word = secondary.words[index]
                let score = calibration[secondary.engine]?.apply(word.confidence) ?? word.confidence
                if let score, score >= 0.75 { words.append(calibrated(word, score)) }
            }
        }
        words.sort { $0.startMs != $1.startMs ? $0.startMs < $1.startMs : $0.endMs < $1.endMs }
        let text = TranscriptSegmentation.normalizedText(words.map(\.text).joined(separator: " "))
        let requested = orderedUnion(primary.context.requestedTerms, secondary.context.requestedTerms)
        let applied = VocabularyEvidence.acousticallySupported(requested, in: words)
        return EngineTranscript(
            engine: "rover",
            model: "\(primary.model)+\(secondary.model)",
            version: "1",
            settings: ["alignment": "timestamped-word-dp", "engines": "\(primary.engine),\(secondary.engine)"],
            text: text,
            language: primary.language ?? secondary.language,
            audioDurationMs: max(primary.audioDurationMs, secondary.audioDurationMs),
            processingDurationMs: primary.processingDurationMs + secondary.processingDurationMs,
            words: words,
            segments: TranscriptSegmentation.segments(from: words),
            diagnostics: ["hypotheses": "2"],
            context: EngineContextEvidence(
                requestedTerms: requested,
                detectedTerms: orderedUnion(primary.context.detectedTerms, secondary.context.detectedTerms),
                appliedTerms: applied
            )
        )
    }

    private static func align(_ left: [TimedWord], _ right: [TimedWord]) -> [Step] {
        let n = left.count, m = right.count
        var cost = Array(repeating: Array(repeating: 0, count: m + 1), count: n + 1)
        for i in 0...n { cost[i][0] = i }
        for j in 0...m { cost[0][j] = j }
        if n > 0, m > 0 {
            for i in 1...n {
                for j in 1...m {
                    let substitution = cost[i - 1][j - 1] + pairCost(left[i - 1], right[j - 1])
                    cost[i][j] = min(substitution, cost[i - 1][j] + 1, cost[i][j - 1] + 1)
                }
            }
        }
        var steps: [Step] = []
        var i = n, j = m
        while i > 0 || j > 0 {
            if i > 0, j > 0 {
                let substitution = pairCost(left[i - 1], right[j - 1])
                if substitution < 2, cost[i][j] == cost[i - 1][j - 1] + substitution {
                    steps.append(.both(i - 1, j - 1)); i -= 1; j -= 1; continue
                }
            }
            if i > 0, cost[i][j] == cost[i - 1][j] + 1 {
                steps.append(.left(i - 1)); i -= 1
            } else {
                steps.append(.right(j - 1)); j -= 1
            }
        }
        return steps.reversed()
    }

    private static func calibrated(_ word: TimedWord, _ confidence: Double?) -> TimedWord {
        TimedWord(text: word.text, startMs: word.startMs, endMs: word.endMs, confidence: confidence)
    }

    private static func sameWord(_ a: String, _ b: String) -> Bool {
        normalize(a) == normalize(b)
    }

    private static func pairCost(_ left: TimedWord, _ right: TimedWord) -> Int {
        let centers = abs((left.startMs + left.endMs) / 2 - (right.startMs + right.endMs) / 2)
        let separated = left.endMs + 500 < right.startMs || right.endMs + 500 < left.startMs
        guard centers <= 1_500 || !separated else { return 2 }
        return sameWord(left.text, right.text) ? 0 : 1
    }

    private static func normalize(_ word: String) -> String {
        word.lowercased().filter { $0.isLetter || $0.isNumber }
    }

    private static func orderedUnion(_ a: [String], _ b: [String]) -> [String] {
        var seen = Set<String>()
        return (a + b).filter { seen.insert($0.lowercased()).inserted }
    }
}

enum TranscriptCalibration {
    static func apply(
        _ transcript: EngineTranscript,
        calibration: [String: QualityProfile.Calibration]
    ) -> EngineTranscript {
        guard let curve = calibration[transcript.engine] else { return transcript }
        let words = transcript.words.map {
            TimedWord(text: $0.text, startMs: $0.startMs, endMs: $0.endMs, confidence: curve.apply($0.confidence))
        }
        let segments = transcript.segments.map { segment in
            TranscriptSegment(
                startMs: segment.startMs,
                endMs: segment.endMs,
                text: segment.text,
                confidence: curve.apply(segment.confidence),
                words: segment.words.map {
                    TimedWord(text: $0.text, startMs: $0.startMs, endMs: $0.endMs, confidence: curve.apply($0.confidence))
                }
            )
        }
        var diagnostics = transcript.diagnostics
        diagnostics["confidence_calibration"] = "held-out-linear-v1"
        return EngineTranscript(
            engine: transcript.engine, model: transcript.model, version: transcript.version,
            settings: transcript.settings, text: transcript.text, language: transcript.language,
            audioDurationMs: transcript.audioDurationMs,
            processingDurationMs: transcript.processingDurationMs,
            words: words, segments: segments, diagnostics: diagnostics, context: transcript.context)
    }
}
